// nats.ws/lib/jetstream/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.consumerOpts = exports.StoreCompression = exports.StorageType = exports.RetentionPolicy = exports.RepublishHeaders = exports.ReplayPolicy = exports.KvWatchInclude = exports.JsHeaders = exports.DiscardPolicy = exports.DirectMsgHeaders = exports.DeliverPolicy = exports.ConsumerEvents = exports.ConsumerDebugEvents = exports.AdvisoryKind = exports.AckPolicy = exports.isHeartbeatMsg = exports.isFlowControlMsg = exports.checkJsError = void 0;
/*
 * Copyright 2023-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var internal_mod_1 = require("./internal_mod");
Object.defineProperty(exports, "checkJsError", { enumerable: true, get: function () { return internal_mod_1.checkJsError; } });
Object.defineProperty(exports, "isFlowControlMsg", { enumerable: true, get: function () { return internal_mod_1.isFlowControlMsg; } });
Object.defineProperty(exports, "isHeartbeatMsg", { enumerable: true, get: function () { return internal_mod_1.isHeartbeatMsg; } });
var internal_mod_2 = require("./internal_mod");
Object.defineProperty(exports, "AckPolicy", { enumerable: true, get: function () { return internal_mod_2.AckPolicy; } });
Object.defineProperty(exports, "AdvisoryKind", { enumerable: true, get: function () { return internal_mod_2.AdvisoryKind; } });
Object.defineProperty(exports, "ConsumerDebugEvents", { enumerable: true, get: function () { return internal_mod_2.ConsumerDebugEvents; } });
Object.defineProperty(exports, "ConsumerEvents", { enumerable: true, get: function () { return internal_mod_2.ConsumerEvents; } });
Object.defineProperty(exports, "DeliverPolicy", { enumerable: true, get: function () { return internal_mod_2.DeliverPolicy; } });
Object.defineProperty(exports, "DirectMsgHeaders", { enumerable: true, get: function () { return internal_mod_2.DirectMsgHeaders; } });
Object.defineProperty(exports, "DiscardPolicy", { enumerable: true, get: function () { return internal_mod_2.DiscardPolicy; } });
Object.defineProperty(exports, "JsHeaders", { enumerable: true, get: function () { return internal_mod_2.JsHeaders; } });
Object.defineProperty(exports, "KvWatchInclude", { enumerable: true, get: function () { return internal_mod_2.KvWatchInclude; } });
Object.defineProperty(exports, "ReplayPolicy", { enumerable: true, get: function () { return internal_mod_2.ReplayPolicy; } });
Object.defineProperty(exports, "RepublishHeaders", { enumerable: true, get: function () { return internal_mod_2.RepublishHeaders; } });
Object.defineProperty(exports, "RetentionPolicy", { enumerable: true, get: function () { return internal_mod_2.RetentionPolicy; } });
Object.defineProperty(exports, "StorageType", { enumerable: true, get: function () { return internal_mod_2.StorageType; } });
Object.defineProperty(exports, "StoreCompression", { enumerable: true, get: function () { return internal_mod_2.StoreCompression; } });
var types_1 = require("./types");
Object.defineProperty(exports, "consumerOpts", { enumerable: true, get: function () { return types_1.consumerOpts; } });
